java/client/src/main/java/glide/connectors/handlers/MessageHandler.java

```diff
@@ -0,0 +1,162 @@
+/** Copyright GLIDE-for-Redis Project Contributors - SPDX Identifier: Apache-2.0 */
+package glide.connectors.handlers;
+
+import static glide.utils.ArrayTransformUtils.castArray;
+
+import glide.api.models.Message;
+import glide.managers.BaseResponseResolver;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentLinkedDeque;
+import java.util.function.BiConsumer;
+import java.util.stream.Stream;
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+import response.ResponseOuterClass.Response;
+
+/** Handler for incoming push messages (subscriptions). */
+@Getter
+@RequiredArgsConstructor
+public class MessageHandler {
+
+    // TODO maybe store `BaseSubscriptionConfiguration` as is?
+    /**
+     * A user callback to call for every incoming message, if given. If missing, messages are pushed
+     * into the {@link #queue}.
+     */
+    private final Optional<BiConsumer<Message, Object>> callback;
+
+    /** An arbitrary user object to be passed to callback. */
+    private final Optional<Object> context;
+
+    /** Helper which extracts data from received {@link Response}s from GLIDE. */
+    private final BaseResponseResolver responseResolver;
+
+    /** A message queue wrapper. */
+    private final MessageQueue queue = new MessageQueue();
+
+    /** Process a push (PUBSUB) message received as a part of {@link Response} from GLIDE. */
+    public void handle(Response response) {
+        Object data = responseResolver.apply(response);
+        if (!(data instanceof Map)) {
+            // TODO log thru error https://github.com/aws/glide-for-redis/pull/1422
+            System.err.println("Received invalid push: empty or in incorrect format.");
+            // TODO do we need to create user a message?
+            return;
+        }
+        @SuppressWarnings("unchecked")
+        Map<String, Object> push = (Map<String, Object>) data;
+        PushKind pushType = Enum.valueOf(PushKind.class, (String) push.get("kind"));
+        String[] values = castArray((Object[]) push.get("values"), String.class);
+        switch (pushType) {
+            case Disconnection:
+                // TODO what to do?
+                break;
+            case PMessage:
+                handle(new Message(values[2], values[1], values[0]));
+                return;
+            case Message:
+            case SMessage:
+                handle(new Message(values[1], values[0]));
+            case Subscribe:
+            case PSubscribe:
+            case SSubscribe:
+            case Unsubscribe:
+            case PUnsubscribe:
+            case SUnsubscribe:
+                // ignore for now
+                break;
+            default:
+                // TODO log thru error https://github.com/aws/glide-for-redis/pull/1422
+                System.err.printf("Received push with unsupported type: %s.\n", pushType);
```

Comment:
https://github.com/Bit-Quill/glide-for-redis/blob/b0a026b5c4ced0202bb33134bef303f447c86218/python/python/glide/redis_client.py#L323-L330